ahl_health_apps_analysis/pipeline/cluster.py

```diff
@@ -0,0 +1,149 @@
+import pandas as pd
+import numpy as np
+import pickle
+import umap.umap_ as umap
+import hdbscan
+import logging
+from sklearn.neighbors import KNeighborsClassifier
+import altair as alt
+import altair_viewer
+from argparse import ArgumentParser
+from sklearn.feature_extraction.text import TfidfTransformer
+from sklearn.feature_extraction.text import TfidfVectorizer
+from sklearn.cluster import KMeans
+
+date = ('2022-11-04')
+details = pd.read_csv(f'outputs/data/preprocessed-description-{date}.csv')
+
+# step 1 - load in word embeddings
+pickle_in = open(f'outputs/data/description-embeddings-{date}.pickle','rb')
+description_embeddings = pickle.load(pickle_in)
+
+# step 2 - reduce word embeddings to lower dimensions
+reducer = umap.UMAP(n_components=2, random_state=1)
+embedding = reducer.fit_transform(description_embeddings)
+
+reducer_clustering = umap.UMAP(n_components=50, random_state=1)
+embedding_clustering = reducer_clustering.fit_transform(description_embeddings)
+
+def get_top_tf_idf_words(response, top_n=2, exclude_words=[]):
+    sorted_nzs = np.argsort(response.data)
+    return [word for word in feature_names[response.indices[sorted_nzs]] if word not in exclude_words][-(top_n+1):-1]
+
+tfidf = TfidfVectorizer(stop_words='english')
+
+if __name__ == '__main__':
+# pick cluster type:
+	def create_argparser():
+
+		parser = ArgumentParser()
+		parser.add_argument('--cluster_type', help="Type of cluster", default="kmeans", type=str)
+		return parser
+
+			
+	parser = create_argparser()
+	args = parser.parse_args()
+
+	if args.cluster_type == "hdbscan":
+
+		# step 3 - 
+		np.random.seed(3667)
+		clusterer = hdbscan.HDBSCAN(
+			min_cluster_size=20,
+			min_samples=1,
+			cluster_selection_method="leaf",
+			prediction_data=True,
+		)
+		clusterer.fit(embedding_clustering)
+		logging.info(
+			f"{len(set(clusterer.labels_))} clusters with {sum([bool(x) for x in clusterer.labels_ if int(x) == -1])} unassigned apps"
+		)
+
+		# step 4 - 
+		df = details.copy()
+		df["x"] = embedding[:, 0]
+		df["y"] = embedding[:, 1]
+		df["cluster"] = [int(x) for x in clusterer.labels_]
+		df["cluster_confidence"] = [str(x) for x in clusterer.probabilities_]
+
+
+		# step 5 - 
```

Comment:
might be worth padding out these "step" comments with a bit more detail `step 5 - assign unassigned apps to their nearest cluster"?